### PR TITLE
Added a new test case for client credentials flow

### PR DESCRIPTION
**Pull request by infohash**
~Addresses the issue with `oic.oauth2.Client.do_access_token_request` for Client Credentials Flow. See #800~

### CODE REVIEW

**Comment by infohash** [comment]
With this fix, if the `request` is `CCAccessTokenRequest`, the `getattr(self, "construct_%s" % request.__name__)` will raise `AttributeError` exception which falls back to `self.construct_request(request, request_args, extra_args)`. We can either implement a new method `construct_CCAccessTokenRequest` or let it fallback to `construct_request` method. Which one do you think is better?

https://github.com/OpenIDC/pyoidc/blob/bd510e49ebd6d8816e3ee94e62a3e86e907a0f8d/src/oic/oauth2/__init__.py#L547-L564

**Comment by schlenk** [comment]
I think i fixed the quality tooling now. Can you please rebase?

**Comment by infohash** [comment]
It's still failing the checks. Can you address it?

**Comment by schlenk** [comment]
Yes. It just fails a check from "black". 

Oh no! 💥 💔 💥
1 file would be reformatted, 108 files would be left unchanged.

**Comment by infohash** [comment]
Reformatted using black. Awaiting for review.

File: src/oic/oauth2/__init__.py [hunk 1]
```diff
@@ -875,14 +880,16 @@ def do_access_token_request(
         state: str = "",
         body_type: ENCODINGS = "json",
         method="POST",
+        request=None,
```
Comment (tpazderka) [inline src/oic/oauth2/__init__.py:883]: The same change should be done in `src/oic/oic/__init__.py` implementation of `Client`.

Also a test of this behaviour would be a nice addition.

Comment (infohash) [reply src/oic/oauth2/__init__.py:883]: Done. Awaiting approval 🙂.

**Comment by infohash** [comment]
It's ready to be merged.

**Comment by tpazderka** [comment]
I am wondering if the `request` argument should actually be there at all. If I am reading this correctly, this can be solved by setting your own `message_factory` with `CCAccessTokenRequest` in the `token_endpoint`. Or am I missing something?

**Comment by infohash** [comment]
It can be done like that, this code was in the test case:

```python
class CCMessageFactory(OauthMessageFactory):
    """We are doing client credentials."""

    token_endpoint = MessageTuple(CCAccessTokenRequest, AccessTokenResponse)

self.client.message_factory = CCMessageFactory
```

We can have this class implemented in the library.

**Comment by infohash** [comment]
@schlenk I have reverted the code but I've addressed the test cases for client credentials which can be merged.

File: tests/test_oauth2.py [hunk 2]
```diff
@@ -420,6 +424,11 @@ def test_request_info_with_req_and_extra_args(self):
         assert h_args == {}
         assert isinstance(cis, AuthorizationRequest)
 
+    def test_request_info_when_request_is_cc_access_token_request(self):
```
Comment (tpazderka) [inline tests/test_oauth2.py:427]: I would like to see some basic verification of the results as well.

**Review by tpazderka — CHANGES_REQUESTED** [review]

File: tests/test_oauth2.py (hunk 2 above)
Comment (infohash) [reply tests/test_oauth2.py:427]: Done. See if this is what you were requesting.

**Comment by infohash** [comment]
@schlenk Can you merge this PR so that test cases in #805 can pass?

**Comment by infohash** [comment]
@tpazderka @schlenk Can you merge this one now? I'll then merge this one in my remaining PRs in the pipeline so they can pass their test cases. I don't want to leave #805 and #806 hanging because there's a draft PR in https://github.com/zamzterz/Flask-pyoidc/pull/132 which depends on them.

I also have a new PR ready which addresses a bug with zero content length response. I'll create it once I clear existing PRs in the pipeline.

**Review by tpazderka — APPROVED** [review]

**Pull request closed without merging**